src/include/platform/DiagnosticDataProvider.h

```diff
@@ -0,0 +1,179 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Defines the public interface for the Device Layer DiagnosticDataProvider object.
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <platform/CHIPDeviceBuildConfig.h>
+#include <platform/GeneralFaults.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+// Maximum length of vendor defined name or prefix of the software thread that is
+// static for the duration of the thread.
+static constexpr size_t kMaxThreadNameLength = 32;
+
+struct ThreadMetrics : public app::Clusters::SoftwareDiagnostics::Structs::ThreadMetrics::Type
+{
+    char NameBuf[kMaxThreadNameLength + 1];
+    ThreadMetrics * Next; /* Pointer to the next structure.  */
+};
+
+/**
+ * Defines the delegate class of Platform Manager to notify platform updates.
+ */
+class DiagnosticsDelegate
+{
+public:
+    virtual ~DiagnosticsDelegate() {}
+
+    /**
+     * @brief
+     *   Called after the current device is rebooted
+     */
+    virtual void OnDeviceRebooted() {}
+};
+
+/**
+ * Provides access to runtime and build-time configuration information for a chip device.
+ */
+class DiagnosticDataProvider
+{
+public:
+    void SetDelegate(DiagnosticsDelegate * delegate) { mDelegate = delegate; }
+    DiagnosticsDelegate * GetDelegate() const { return mDelegate; }
+
+    /**
+     * General Diagnostics methods.
+     */
+    virtual CHIP_ERROR GetRebootCount(uint16_t & rebootCount);
+    virtual CHIP_ERROR GetUpTime(uint64_t & upTime);
+    virtual CHIP_ERROR GetTotalOperationalHours(uint32_t & totalOperationalHours);
+    virtual CHIP_ERROR GetBootReason(uint8_t & bootReason);
+    virtual CHIP_ERROR GetActiveHardwareFaults(GeneralFaults<kMaxHardwareFaults> & hardwareFaults);
+    virtual CHIP_ERROR GetActiveRadioFaults(GeneralFaults<kMaxRadioFaults> & radioFaults);
+    virtual CHIP_ERROR GetActiveNetworkFaults(GeneralFaults<kMaxNetworkFaults> & networkFaults);
+
+    /**
+     * Software Diagnostics methods.
+     */
+    virtual CHIP_ERROR GetCurrentHeapFree(uint64_t & currentHeapFree);
+    virtual CHIP_ERROR GetCurrentHeapUsed(uint64_t & currentHeapUsed);
+    virtual CHIP_ERROR GetCurrentHeapHighWatermark(uint64_t & currentHeapHighWatermark);
+
+    /*
+     * Get the linked list of thread metrics of the current plaform. After usage, each caller of GetThreadMetrics
+     * needs to release the thread metrics list it gets via ReleaseThreadMetrics.
+     *
+     */
+    virtual CHIP_ERROR GetThreadMetrics(ThreadMetrics ** threadMetricsOut);
```

Comment:
`ForEachThread(void (*visitor)(const ThreadInfo&, uintptr_t), uintptr_t context)` might be more scalable interface as some applications may use quite a lot of threads.

Reply:
I think this also apply to interface    
```
CHIP_ERROR GetNetworkInterfaces(NetworkInterface ** netifpp);
void ReleaseNetworkInterfaces(NetworkInterface * netifp);
```

What I would like to do is to add iterator to netifpp struct, so we can just use following syntax to iterate all available active interface.  

```
        for (auto & ifp : netifs)
        {

        }

```
I will do it in follow-up https://github.com/project-chip/connectedhomeip/issues/12106 and leave this PR only to refactor the current implemenation from PlatformMgr to DiagnosticDataProvider

Reply:
Yeah, a custom iterator would also be nice. In case of network interfaces we already have `InterfaceIterator` in `InetInterface.h` which is implemented for most platforms, so it's worth considering using it instead of creating another mechanism.

Reply:
But that's just an idea. I'm not requesting these changes as part of this PR.